ddtrace/contrib/aiohttp/middlewares.py

```diff
@@ -24,106 +46,141 @@ async def trace_middleware(app, handler):
     if trace_query_string is None:
         trace_query_string = config.http.trace_query_string
     if trace_query_string:
         request_span.set_tag(http.QUERY_STRING, request.query_string)
 
-    trace_utils.set_http_meta(
-        request_span,
-        config.aiohttp,
-        method=request.method,
-        url=str(request.url),  # DEV: request.url is a yarl's URL object
-        status_code=response.status,
-        request_headers=request.headers,
-        response_headers=response.headers,
-    )
+    if not iswrapped(request._payload) and isinstance(request._payload, aiohttp.streams.StreamReader):
+        tags = {
+            tag: request_span.get_tag(tag)
+            for tag in {http.URL, http.METHOD, http.QUERY_STRING}
+            if request_span.get_tag(tag)
+        }
+
+        pin = Pin(service, tracer=tracer, tags=tags, _config=config.aiohttp_server)
+        request._payload = _WrappedStreamReader(request._payload, pin, request_span.trace_id, request_span.span_id,
+                                                request_span.resource)
+
+    try:
+        response = await handler(request)
+        return response
+    except Exception:
+        request_span.set_traceback()
+        raise
+
+
+async def trace_middleware_1x(app, handler):
+    """
+    ``aiohttp`` middleware that traces the handler execution.
+    Because handlers are run in different tasks for each request, we attach the Context
+    instance both to the Task and to the Request objects. In this way:
+
+    * the Task is used by the internal automatic instrumentation
+    * the ``Context`` attached to the request can be freely used in the application code
+    """
+    return functools.partial(trace_middleware_2x, handler=handler, app=app)
+
+
+trace_middleware = trace_middleware_2x if AIOHTTP_HAS_MIDDLEWARE else trace_middleware_1x
+
+
+async def on_prepare(request, response):
+    """
+    The on_prepare signal is used to close the request span that is created during
+    the trace middleware execution.
+    """
+    # safe-guard: discard if we don't have a request span
+    request_span = request.get(REQUEST_SPAN_KEY, None)
+    if not request_span:
+        return
+
+    # default resource name
+    if not request_span.resource:
+        request_span.resource = stringify(response.status)
+
+    request_span.set_tag(http.STATUS_CODE, response.status)
+
+    # TOOD: normally we'd do this but we want to pin the tags to the stream response via middleware
+    # trace_utils.set_http_meta(
+    #     request_span,
+    #     config.aiohttp,
+    #     method=request.method,
+    #     url=str(request.url),  # DEV: request.url is a yarl's URL object
+    #     status_code=response.status,
+    #     request_headers=request.headers,
+    #     response_headers=response.headers,
+    # )
 
     request_span.finish()
 
 
-def trace_app(app, tracer, service="aiohttp-web"):
+def trace_app(app, tracer, service="aiohttp-web", distributed_tracing=True):
```

Comment:
we can remove this param, and have callers set `app[CONFIG_KEY]['distributed_tracing_enabled'] = True`